storage/storage.py

```diff
@@ -129,6 +129,10 @@ def get_download_url(self, seconds: int = 60, key: Optional[str] = None) -> str:
         raise NotImplementedError(
             "{} does not implement 'get_download_url'".format(self._class_name()))
 
+    def get_sanitized_uri(self) -> str:
+        raise NotImplementedError(
```

Comment:
@spiralman @shroudofthurin I think I'm okay with the default implementation being in the base class and even making it overly restrictive.

One thing we could do is have the base class take an optional whitelist of identifiers indicating URI parts that are safe to include in the sanitized URI. Then, subclasses that want to include the query parameters and username can tell the base to include them. e.g.

```python
class DerivedStorage(Storage):
    def __init__(self):
         super().__init__(sanitize_whitelist=[USERNAME, QUERY])
```

...could result in storage URIs that look like: `proto://username@netloc/path?query=parameters`

Reply:
@tjensen @spiralman @shroudofthurin While I agree that the current implementations are very similar, this is explicit boilerplate I believe to be a Good Thing[tm], since it's such an important consideration for any project using storage URIs. Thus my preference would be that it should raise, as it does in the code now (although I might prefer renaming the "sanitized_uri" helper to "sanitize_basic_auth" or something more specific.) If we do use a default behavior instead of raising, I recommend that we at least log a warning when a storage class with the "default" sterilization behavior is imported / registered.

Reply:
I definitely agree with renaming the `sanitize_uri` helper to something more explicit (basic_auth feels a little confusing, because we're not actually using it for HTTP Basic Auth). The spec calls it "user information", so maybe `remove_user_info`?

Anyway, I'm 50/50 on whether there should be a default implementation or it should raise, so it seems like we have a hung jury. :-) I guess I'd recommend @shroudofthurin make the deciding vote, since it's her PR.

Reply:
@spiralman I think I will go with your original solution:

`One solution to that would be to have the default implementation be overly restrictive (say, only returning the scheme, hostname, port and path, no query parameters, etc), and then have implementations override it which want to include extra values in the URL.`